### PR TITLE
7.0 add sale pricelist restrict product in so

### PR DESCRIPTION
**Pull request by veloutin**
add the sale_pricelist_restrict_product_in_so module, which, when installed,
restricts products in sale order lines to those allowed by the selected pricelist.

### CODE REVIEW

**Comment by veloutin** [comment]
@pedrobaeza Those errors do not seem in any way related to the new module...
Same problem as with #17 ?

<pre>
openerp_test openerp.addons.base.ir.ir_ui_view: Can't render view sale_validity.view_order_form_validity for model: sale.order
Traceback (most recent call last):
  File "/home/travis/build/OCA/odoo-7.0/openerp/addons/base/ir/ir_ui_view.py", line 126, in _check_render_view
    fvg = self.pool.get(view.model).fields_view_get(cr, uid, view_id=view.id, view_type=view.type, context=context)
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 2263, in fields_view_get
    arch=apply_view_inheritance(cr, user, source, sql_res['id']),
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 2221, in apply_view_inheritance
    source = apply_inheritance_specs(source, view_arch, view_id)
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 2205, in apply_inheritance_specs
    raise_view_error("Element '%s' not found in parent view '%%(parent_xml_id)s'" % tag, inherit_id)
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 2098, in raise_view_error
    %  (child_view.xml_id, self._name, error_msg))
AttributeError: View definition error for inherited view 'sale_stock.view_order_form_inherit' on model 'sale.order': Element '<xpath expr="//button[@name='action_cancel']">' not found in parent view 'sale.view_order_form'
openerp_test openerp.tools.convert: Parse error in /home/travis/build/OCA/sale-workflow/sale_validity/view/sale_order.xml:5: 
<record id="view_order_form_validity" model="ir.ui.view">
        <field name="name">sale.order.form.validity</field>
        <field name="model">sale.order</field>
        <field name="inherit_id" ref="sale.view_order_form"/>
        <field name="arch" type="xml">
            <data><field name="date_order" position="after">
                <field name="date_validity"/>
            </field>
            <field name="date_order" position="attributes">
                <attribute name="on_change">date_order_change(date_order, date_validity, company_id, context)</attribute>
            </field>
        </data></field>
    </record>
Traceback (most recent call last):
  File "/home/travis/build/OCA/odoo-7.0/openerp/tools/convert.py", line 852, in parse
    self._tags[rec.tag](self.cr, rec, n)
  File "/home/travis/build/OCA/odoo-7.0/openerp/tools/convert.py", line 819, in _tag_record
    id = self.pool.get('ir.model.data')._update(cr, self.uid, rec_model, self.module, res, rec_id or False, not self.isnoupdate(data_node), noupdate=self.isnoupdate(data_node), mode=self.mode, context=rec_context )
  File "/home/travis/build/OCA/odoo-7.0/openerp/addons/base/ir/ir_model.py", line 952, in _update
    model_obj.write(cr, uid, [res_id], values, context=context)
  File "/home/travis/build/OCA/odoo-7.0/openerp/addons/base/ir/ir_ui_view.py", line 214, in write
    return super(view, self).write(cr, uid, ids, vals, context)
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 4298, in write
    self._validate(cr, user, ids, context)
  File "/home/travis/build/OCA/odoo-7.0/openerp/osv/orm.py", line 1572, in _validate
    raise except_orm('ValidateError', '\n'.join(error_msgs))
except_orm: ('ValidateError', u'Error occurred while validating the field(s) arch: Invalid XML for View Architecture!')
</pre>

**Comment by mcassuto** [comment]
@pedrobaeza : I tested the module sale_validity on a fresh DB and I successfully installed it without any problem.... Pedro please advise, since the module that makes the test to fail is not the one veloutin wants to pull

**Comment by yvaucher** [comment]
@mcassuto I was certainly due to issue https://github.com/OCA/sale-workflow/issues/13 
This has been fixed. However, travis still has to be fixed as some tests are red.

However, by rebasing, we should get a proper state of your PR in travis

So please rebase.

**Comment by veloutin** [comment]
Rebased on OCA 7.0 branch

**Comment by bwrsandman** [comment]
Hey @veloutin, as a rule, please put one model/class per python file.
- product.py or product_product.py
- product_pricelist.py
- product_pricelist_item.py

It makes maintenance and review easier.

**Comment by mcassuto** [comment]
@yvaucher  @bwrsandman : refactor done !
@veloutin please complete if Travis still fails

**Comment by yvaucher** [comment]
Here is a PEP8 error to fix

./sale_pricelist_restrict_product_in_so/product.py:44:32: F821 undefined name 'ANY_PRODUCT'

Plus, there is an error in tests:
module sale_pricelist_restrict_product_in_so: at least one error occurred in a tests

**Comment by veloutin** [comment]
Fixed the pep8 error and completed @mcassuto 's refactor.

**Comment by veloutin** [comment]
Travis is failing in the sale_delivery_term module yaml tests, when calling check_master_line.
 @eLBati , do you have an idea how to fix this?

**Comment by bwrsandman** [comment]
@veloutin half fixed in #22

**Comment by eLBati** [comment]
Yes, `sale_delivery_term` error about `Delivery term lines are 4` is fixed with https://github.com/OCA/sale-workflow/pull/22
The `sale_exception_nostock` error saying `Incorrect state manual instead of draft`, that only happens with OCB, is still open

File: sale_pricelist_restrict_product_in_so/product.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class ProductProduct(orm.Model):
+    _inherit = 'product.product'
+
+    def search(self, cr, uid, args,
+               offset=0, limit=None,
+               order=None, context=None, count=False):
+        if context is None:
+            context = {}
+        if context and "pricelist" in context:
```
Comment (pedrobaeza) [inline sale_pricelist_restrict_product_in_so/product.py]: Better

```
if context.get('pricelist'):
```

**Comment by mcassuto** [comment]
@pedrobaeza : hI Pedro, 
if context.get('pricelist'):   fixed !

@yvaucher: Are there other fixings to do ?

**Comment by yvaucher** [comment]
@mcassuto LGTM, thanks

:+1:

**Comment by bwrsandman** [comment]
:+1:

**Pull request merged**